src/newrelic/processor.go

```diff
@@ -6,819 +6,817 @@
+                r.ConnectTimestamp = uint64(app.connectTime.Unix())
+                r.HarvestFrequency = uint16(app.harvestFrequency.Seconds())
+                r.SamplingTarget = uint16(app.samplingTarget)
+            }
+        }
+        // Send the response back before attempting to connect the application
+        m.ResultChan <- r
+        if nil != app {
+            p.considerConnect(app)
+        }
+    }()
+
+    if nil != m.ID {
+        if _, ok := p.harvests[*m.ID]; ok {
+            r.RunIDValid = true
+            return
+        }
+        // This agent run id must be out of date, fall through:
+    }
+
+    key := m.Info.Key()
+    app = p.apps[key]
+    if nil != app {
+        //set LastActivity so we treat an AppInfo request for
+        //a known app as activity.
+        app.LastActivity = time.Now()
+        return
+    }
+
+    if len(p.apps) > limits.AppLimit {
+        log.Errorf("unable to add app '%s', limit of %d applications reached",
+        m.Info, limits.AppLimit)
+        return
+    }
+
+    app = NewApp(m.Info)
+    p.apps[key] = app
+}
+
+func processConnectMessages(reply collector.RPMResponse) {
+    var msgs struct {
+        Messages []struct {
+            Message string `json:"message"`
+            Level   string `json:"level"`
+        } `json:"messages"`
+    }
+
+    err := json.Unmarshal(reply.Body, &msgs)
+    if nil != err {
+        return
+    }
+
+    for _, msg := range msgs.Messages {
+        switch strings.ToLower(msg.Level) {
+        case "error":
+            log.Errorf("%s", msg.Message)
+        case "warn":
+            log.Warnf("%s", msg.Message)
+        case "info":
+            log.Infof("%s", msg.Message)
+        case "debug", "verbose":
+            log.Debugf("%s", msg.Message)
+        }
+    }
 }
 
 func (p *Processor) processConnectAttempt(rep ConnectAttempt) {
-	app := p.apps[rep.Key]
-	if nil == app {
-		return
-	}
-
-	if nil != rep.Err {
-		switch {
-		case collector.IsDisconnect(rep.Err):
-			app.state = AppStateDisconnected
-		case collector.IsLicenseException(rep.Err):
-			app.state = AppStateInvalidLicense
-		default:
-			// Try again later.
-			app.state = AppStateUnknown
-		}
-
+    app := p.apps[rep.Key]
+    if nil == app {
+        return
+    }
+
+    app.RawConnectReply = rep.RawReply.Body
+    if  rep.RawReply.IsDisconnect() {
+        app.state = AppStateDisconnected
+        log.Warnf("app '%s' connect attempt returned %s", app, rep.RawReply.Err)
+        return
+    }  else if rep.RawReply.IsRestartException() {
+        app.state = AppStateRestart
+        // in accord with the spec, invalid license is a restart exception. Except we want
+        //    to shutdown instead of restart.
+        if rep.RawReply.IsInvalidLicense() {
+            app.state = AppStateInvalidLicense
+        }
+        log.Warnf("app '%s' connect attempt returned %s", app, rep.RawReply.Err)
+        return
+    } else if nil != rep.Err {
```

Comment:
Can we move the check for an error to the top so we don't go through several if checks that we know will fail?